source/opt/def_use_manager.h

```diff
@@ -84,27 +87,50 @@ class DefUseManager {
   // information kept in this manager, but not the operands in the original
   // instructions.
   bool KillDef(uint32_t id);
+  // Turns the given instruction |inst| to a Nop, also erases both the use
+  // records of the result id of |inst| (if any) and the corresponding use
+  // records of: |inst| using |inst|'s operand ids.
+  void KillInst(ir::Instruction* inst);
   // Replaces all uses of |before| id with |after| id. Returns true if any
   // replacement happens. This method does not kill the definition of the
   // |before| id. If |after| is the same as |before|, does nothing and returns
   // false.
   bool ReplaceAllUsesWith(uint32_t before, uint32_t after);
 
  private:
-  // Clear the internal def-use record of a defined id if the given |def_id| is
-  // recorded by this manager. This method will erase both the uses of |def_id|
-  // and the |def_id|-generating instruction's use information kept in this
-  // manager, but not the operands in the original instructions.
+  // Clears the internal def-use records of a defined id if the given |def_id|
+  // is recorded by this manager. Does nothing if |def_id| has not been
+  // recorded yet. This method will erase both the uses of |def_id| and the
+  // |def_id|-generating instruction's use information kept in this manager,
+  // but not the operands in the original instructions.
   void ClearDef(uint32_t def_id);
 
-  using ResultIdToUsedIdsMap =
-      std::unordered_map<uint32_t, std::vector<uint32_t>>;
-
+  // Clears the internal def-use records of the given instruction |inst| if it
+  // has been analyzed by this manager. The use information of its operand ids
+  // will be updated: "the record: |inst| uses |operand id| will be removed".
+  // If the instruction is defining a result id, the uses of the result id will
+  // also be removed. Note that if |inst| has not been analyzed before, this
+  // function does nothing even though |inst| may define an existing result id.
+  void ClearInst(ir::Instruction* inst);
+
+  // Returns true if the operand's id should be recorded, otherwise returns
+  // false;
+  bool ShouldRecord(const ir::Operand& operand) const;
+
+  // Erases the record that: instruction |user| uses id |used_id|.
+  void EraseInstUseIdRecord(const ir::Instruction& user, uint32_t used_id);
+
+  // Resets the internal records
+  void Reset() {
+    analyzed_insts_.clear();
+    id_to_def_.clear();
+    id_to_uses_.clear();
+  }
+
+  std::unordered_set<ir::Instruction*>
+      analyzed_insts_;      // Analyzed instructions
```

Comment:
There may be four cases (depends on whether we want to make ir::Instruction to be immutable later) we need to differentiate:
1) a new instruction without result_id but with already analyzed operands.
2) a new instruction with result_id and also has its operands analyzed.
3) an analyzed but mutated instruction, the instruction does have a result id.
4) an analyzed but mutated instruction, the instruction does not have a result id.

For case 1), the desired result is that, the def-use records of the analyzed operands should be +1 to have the new added instruction as their new uses.

For case 2), the desired result is that, the original result_id defining instruction should be cleared first, then the new added instruction can update the records for the involved operands. This is a -1 followed by +1 for the operands.

For case 3), we should -1 first then +1, same as a new instruction with existing result id.

For case 4), we should also -1 first then +1, same as a new instruction with existing result id, but this is different with case 1).

To differentiate the four cases we have two options:

a) See if the instruction has a result id.
If the instruction has a result id, clear internal records for the id, then analyze the new instruction. Otherwise analyze the new instruction directly. But this way we can not differentiate case 1 and case 4. We will -1 for the operands in case 1.

b) Use a set to hold all the analyzed instructions.
If the instruction has been analyzed, clear its record first.
If the result id has been analyzed before, clear its def_id record also.
Add record for the given instruction as if it is a new instruction.

Actually the fundamental solution to this hassle is to use instruction* instead of id(uint32_t) for the internal maps then we can unify the behavior. But that may changed the structure a lot. This is the reason why we need such a set here.